README.md
# Stacker

Simple web app to facilitate stack based video meetings.

## Dev

This project uses [`stack`](https://docs.haskellstack.org/en/stable/README/) for builds
and deployments. Use the classic

```sh
stack clean
stack build
stack exec stacker
```

to get clean, build, then run locally.

## Deploy to Heroku

I am currently hosting my instance of `stacker` on Heroku. To do the same you
will need to add a `stack` buildpack for your project. [This
one](https://github.com/mfine/heroku-buildpack-stack) has worked well for me.

### Heroku Free Tier

*NOTE* if you are on the Heroku free tier consider using the environment variable
`KEEPALIVE_URL` to create a new keepalive page in the application. This page is
useful to keep the application from being put to sleep after ~30 minutes when
web requests are no longer coming in but meetings are still ongoing. For example,
adding the variable `KEEAPLIVE_URL=/keepalive` will add a keepalive endpoint that
just refreshes the page every 10 minutes to keep the application running.

Check out an instance of the free tier at at https://chat-stacker.herokuapp.com


## Docker

### Build

Build this image in the usual way:

```
docker build . -t stacker:<tag name>
```

### Run

Run the image with:

```
docker run --rm --name stacker -p 80:80 stacker:<tag name>
```

Note, if your hosting provider expects a specific port or host you can assign it
with:

```
docker run -e PORT=<new port> -e HOST=<new host> --rm ...  
```


## TODO

- ~Dockerize~
- ~Add support for stacking on behalf of other attendees~
- Add progressive stack / ~method to move people around the stack~
- Add support for context based stacking
- ~Add speak counter to attendees~
- Prettify UI
- Consider separating the websocket app from the http app


CHANGELOG.md
# Changelog

All notable changes to this project will be documented here.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.5.0]
### Added
- Support for adding other people to meeting. Useful when facilitating users who
  cannot use stacker easily, like phone users
- Keepalive url for Heroku free tier

## [0.4.0] - 2021-05-18
### Added
- More docs
- Dockerized
- Supports stacking for others
- Added support for moving position in stack
- Added links to policies in footer
- Added support for canceling a stack
- Added speak counter
### Fix
- Added client-side form validation to disallow blank room or user names
- Handles duplicate usernames

## [0.3.0] - 2021-05-14
### Added
- More sensible stack update when creating new local stack from broad stack
- Added missing cookie deletion functionality

## [0.2.1] - 2021-05-14
### Fix
- Delete user also clears removes the user from local and global stacks

## [0.2.0] - 2021-05-14
### Added
- Better support for clients rejoining rooms (properly refreshes / retrieves room data)
- Support for attendees leaving rooms
- Support for closing rooms

##  [0.1.0] - 2021-05-13
### Added
- Initial implementation of `stacker`. Includes:
-- Basic functionality to support 2 simultaneous stacks
-- Supports single clients/browsers using multiple different rooms
-- Easy sharing


Data/Queue.hs
{-# Language DerivingVia #-}
{-# Language StandaloneDeriving #-}
module Data.Queue
  (
    Queue
  , empty
  , fromList
  , next
  , remove
  , peak
  , depth
  , sort
  , (|>)
  ) where
import           Protolude            hiding (empty, on, sort)
import qualified Data.Sequence as Seq

newtype Queue a = Q (Seq a) deriving (Eq, Ord, Read, Show)
deriving via Seq instance Foldable Queue

empty :: Queue a
empty = Q Seq.empty

fromList :: [a] -> Queue a
fromList = Q . Seq.fromList

(|>) :: Queue a -> a -> Queue a
q |> x = on (Seq.|> x) q

next :: Queue a -> Queue a
next = on (Seq.drop 1)

remove :: Eq a => a -> Queue a -> Queue a
remove el = on (Seq.filter (/= el))

peak :: Queue a -> Maybe a
peak (Q s) = Seq.lookup 0 s

depth :: Queue a -> Int
depth (Q s) = Seq.length s

sort :: Ord a => Queue a -> Queue a
sort (Q s) = Q (Seq.unstableSort s)

on :: (Seq a -> Seq a) -> Queue a -> Queue a
on f (Q s) = Q $ f s


Data/Types.hs
{-# LANGUAGE DeriveGeneric       #-}
{-# LANGUAGE NamedFieldPuns      #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Data.Types
  (
    RoomID(..)
  , UserID(..)
  , RoomState'(..)
  , RoomState
  , SpeakReq
  , StackType(..)
  , Payload(..)
  , newRoom
  , nilRoom
  , qGlobal
  , qLocal
  , q
  , deleteUser
  , newUser
  , nextLocal
  , nextGlobal
  , lengthLocal
  , reorder
  , cancelStack
  , peakNextGlobal
  , userList
  , uuidToText
  , textToUUID
  , textToRoomID
  , roomIDToText
  , unRoomID
  ) where
import           Protolude                   hiding (Text, empty, local, sort)
import           Data.Aeson
import           Data.Aeson.Encoding (lazyText)
import qualified Data.Map.Strict     as Map
import           Data.Text.Lazy      (Text)
import qualified Data.Text.Lazy      as Text
import           Data.UUID           (UUID)
import qualified Data.UUID           as UUID
import           GHC.Generics        ()

import           Data.Queue          as Q

newtype RoomID   = RoomID UUID deriving (Eq, Generic, Ord, Read, Show)
newtype UserID   = UserID Text deriving (Eq, Generic, Ord, Read, Show)

unRoomID :: RoomID -> UUID
unRoomID (RoomID u) = u

uuidToText :: UUID -> Text
uuidToText = Text.fromStrict . UUID.toText

textToUUID :: Text -> Maybe UUID
textToUUID = UUID.fromText . Text.toStrict

textToRoomID :: Text -> Maybe RoomID
textToRoomID = fmap RoomID . textToUUID

roomIDToText :: RoomID -> Text
roomIDToText = uuidToText . unRoomID

type SpeakReq = UserID
type RoomName = Text

data RoomState' a = RoomState {
    roomID    :: RoomID,
    roomName  :: RoomName,
    users     :: Map UserID Int,
    rglobal   :: Queue a,
    rlocal    :: Queue a
  } deriving (Eq, Generic, Ord, Read, Show)

type RoomState = RoomState' SpeakReq

userList :: RoomState' a -> [UserID]
userList = Map.keys . users

newRoom :: RoomID -> RoomName -> RoomState' a
newRoom rid rname = RoomState rid rname Map.empty Q.empty Q.empty

nilRoom :: RoomState' a
nilRoom = newRoom (RoomID UUID.nil) ""

-- Returns Nothing when username is taken
newUser :: UserID -> RoomState' a -> Maybe (RoomState' a)
newUser user roomState@RoomState{users} =
  if Map.member user users
  then Nothing
  else Just roomState{users = Map.insert user 0 users}

deleteUser :: UserID -> RoomState -> RoomState
deleteUser user roomState@RoomState{users, rglobal, rlocal} = roomState{
    users = Map.delete user users,
    rglobal = Q.remove user rglobal,
    rlocal  = Q.remove user rlocal
  }

qLocal :: a -> RoomState' a -> RoomState' a
qLocal req = ql (|> req)

qGlobal :: a -> RoomState' a -> RoomState' a
qGlobal req = qg (|> req)

q :: StackType -> a -> RoomState' a -> RoomState' a
q = stackToFunc qLocal qGlobal

nextLocal :: RoomState -> RoomState
-- nextLocal = ql next where
nextLocal rstate@RoomState{rlocal, users} = rstate{rlocal=next rlocal, users=users'} where
  users' = case Q.peak rlocal of
             Nothing -> users
             Just u  -> Map.adjust (+1) u users

nextGlobal :: RoomState' a -> RoomState' a
nextGlobal = qg next

lengthLocal :: RoomState' a -> Int
lengthLocal = Q.depth . rlocal

peakNextGlobal :: RoomState' a -> Maybe a
peakNextGlobal = Q.peak . rglobal

reorder :: forall a. Ord a => [a] -> StackType -> RoomState' a -> RoomState' a
reorder suggestedList = stackToFunc (ql replaceQ) (qg replaceQ) where
  replaceQ :: Queue a -> Queue a
  replaceQ existingQueue =
    -- Checks that the existing and suggested queue at least contain
    -- the same elements
    if Q.sort suggestedQueue == Q.sort existingQueue
    then suggestedQueue
    else existingQueue
  suggestedQueue :: Queue a
  suggestedQueue = Q.fromList suggestedList

cancelStack :: forall a. Ord a => [a] -> StackType -> RoomState' a -> RoomState' a
cancelStack suggestedList = stackToFunc (ql replaceQ) (qg replaceQ) where
  replaceQ :: Queue a -> Queue a
  replaceQ existingQueue =
    -- Checks that the existing and suggested queue at least contain
    -- the same elements
    if Q.depth suggestedQueue == Q.depth existingQueue - 1
    then suggestedQueue
    else existingQueue
  suggestedQueue :: Queue a
  suggestedQueue = Q.fromList suggestedList

ql :: (Queue a -> Queue a) -> RoomState' a -> RoomState' a
ql queueFunc roomState@RoomState{rlocal} = roomState{rlocal=queueFunc rlocal}

qg :: (Queue a -> Queue a) -> RoomState' a -> RoomState' a
qg queueFunc roomState@RoomState{rglobal} = roomState{rglobal=queueFunc rglobal}

stackToFunc :: (a -> b) -> (a -> b) -> StackType -> (a -> b)
stackToFunc local _ LOCAL = local
stackToFunc _ broad BROAD = broad

data StackType = LOCAL | BROAD deriving (Eq, Generic, Ord, Read, Show)

data Payload =
  QUEUE   { stack :: StackType                                 } |
  QOTHER  { stack :: StackType, other :: UserID                } |
  NEXT    { stack :: StackType                                 } |
  LEAVE                                                          |
  CLOSE                                                          |
  REORDER { stack :: StackType, newstack :: [UserID]           } |
  CANCEL  { stack :: StackType, newstack :: [UserID]           } |
  ADD   { other :: UserID                                      } |
  WORLD { attendeeToCount :: Map UserID Int, local :: [UserID],
          broad :: [UserID], name :: Text }                      |
  CLIENTINIT { attendee :: UserID, room :: RoomID              } |
  DUPLICATEUSER { attendees :: [UserID]                        }
  deriving (Eq, Generic, Ord, Read, Show)


instance FromJSON    UserID
instance ToJSON      UserID
instance ToJSONKey   UserID where
  toJSONKey = ToJSONKeyText f g where
    f (UserID u) = Text.toStrict u
    g (UserID u) = lazyText u
instance FromJSONKey UserID
instance FromJSON    RoomID
instance ToJSON      RoomID
instance FromJSON    StackType
instance ToJSON      StackType
instance ToJSON      Payload where
  toJSON    = genericToJSON    defaultOptions { sumEncoding = TaggedObject "action" "" }
instance FromJSON Payload where
  parseJSON = genericParseJSON defaultOptions { sumEncoding = TaggedObject "action" "" }


static/client.js
var debug = false;
const debugLog = function(msg) {
    if(debug) console.log(msg);
};

const roomid     = location.pathname.split('/')[2];
const cookieName = `username-${roomid}`;
const roomCookie = document.cookie.split("; ")
      .find(cookie => cookie.startsWith(cookieName));
var username = roomCookie.split('=')[1];
const attendees   = {};
const localStack  = [];
const globalStack = [];

let scheme = "wss";
let port   = 443;
if (location.protocol === 'http:') {
    scheme = "ws";
    port   = 80;
}
const ws = new WebSocket(`${scheme}://${location.host}${location.pathname}:${port}`);

ws.onclose = function(m) {
    console.log(`Connection close reason: ${m.reason}`);
    alert("This room is now closed");
    window.location = '/';
};

window.onbeforeunload = evt => {
    ws.close();
};

const LOCALSTACK = 'LOCAL';
const BROADSTACK = 'BROAD';
const protocol = function(action, extra = {}) {
    return (el) => {
        const stacktype = typeof el === 'string' ? el : el.parentElement.id;
        const msg = Object.assign(extra, { action: action, stack: stacktype });
        debugLog(`From protocol: [${JSON.stringify(msg)}]`);
        ws.send(JSON.stringify(msg));
    };
};
const QUEUE_ACT     = 'QUEUE';
const NEXT_ACT      = 'NEXT';
const QOTHER_ACT    = 'QOTHER';
const ATTENDEES_ACT = 'UPDATE_ATTENDEES';
const UPDATE_WORLD  = 'WORLD';
const REORDER_ACT   = 'REORDER';
const CANCEL_ACT    = 'CANCEL';
const ADD_ACT       = 'ADD';
const q      = protocol(QUEUE_ACT);
const next   = protocol(NEXT_ACT);
const qother = function(el) {
    const selID    = `${el.parentElement.id.toLowerCase()}-stack-other`;
    const attendee = document.getElementById(selID).value;
    protocol(QOTHER_ACT, {other: attendee})(el);
};

const shiftUp = function(i, stacktype) {
    shifter(i, i - 1, stacktype);
};

const shiftDown = function(i, stacktype) {
    shifter(i, i + 1, stacktype);
};

const swap = function(xs, l, r) {
    if (l < 0 || l >= xs.length || r < 0 || r >= xs.length) return xs;
    const holdthis = xs[l];
    xs[l] = xs[r];
    xs[r] = holdthis;
    return xs;
};

const shifter = function(l, r, stacktype) {
    const stack = [...stacktype === LOCALSTACK ? localStack : globalStack];
    swap(stack, l, r);
    protocol(REORDER_ACT, { newstack: stack })(stacktype);
};

const cancel = function(index, stacktype) {
    const stack = [...stacktype === LOCALSTACK ? localStack : globalStack];
    stack.splice(index, 1);
    protocol(CANCEL_ACT, { newstack: stack })(stacktype);
};

const addOtherAttendee = function() {
    const otherEl = document.getElementById("other-attendee-username");
    const other   = otherEl.value.trim();
    if (!other) {
        alert("Attendee name is required!");
        return;
    }
    const msg = { action: ADD_ACT, other: other };
    debugLog(`From protocol: [${JSON.stringify(msg)}]`);
    ws.send(JSON.stringify(msg));
    otherEl.value = '';
};

const updateAttendees = function(attendees) {
    const attendeesel = document.getElementById('attendees');
    const selectl     = document.getElementById('local-stack-other');
    const selectb     = document.getElementById('broad-stack-other');
    attendeesel.innerHTML = '';
    selectl.innerHTML     = '';
    selectb.innerHTML     = '';
    for(const [attendee, count] of Object.entries(attendees)) {
        const row = document.createElement('tr');
        const attendeeEl = document.createElement('td');
        attendeeEl.appendChild(document.createTextNode(attendee));
        row.appendChild(attendeeEl);
        const countEl = document.createElement('td');
        countEl.appendChild(document.createTextNode(count));
        row.appendChild(countEl);
        attendeesel.appendChild(row);
        // These fill in drop downs of only other attendees
        if (attendee !== username) {
            const localopt = document.createElement('option');
            localopt.value = attendee;
            localopt.appendChild(document.createTextNode(attendee));
            selectl.appendChild(localopt);
            const broadopt = document.createElement('option');
            broadopt.value = attendee;
            broadopt.appendChild(document.createTextNode(attendee));
            selectb.appendChild(broadopt);
        }
    };
};

const newQueue = function(stacktype, items) {
    const stackel = document.getElementById(`${stacktype.toLowerCase()}-discussion`);
    stackel.innerHTML = '';
    let i = 0;
    items.forEach(item => {
        let k = i;
        const newli   = document.createElement('li');
        newli.appendChild(document.createTextNode(item));
        const upButton = document.createElement('button');
        upButton.appendChild(document.createTextNode('↑'));
        upButton.onclick = function() { shiftUp(k, stacktype); };
        const downButton = document.createElement('button');
        downButton.appendChild(document.createTextNode('↓'));
        downButton.onclick = function() { shiftDown(k, stacktype); };

        const cancelButton = document.createElement('button');
        cancelButton.appendChild(document.createTextNode("✕"));
        cancelButton.onclick = function() { cancel(k, stacktype); };

        newli.appendChild(upButton);
        newli.appendChild(downButton);
        newli.appendChild(cancelButton);
        stackel.appendChild(newli);
        i++;
    });
};

const setRoomName = function(name) {
    document.getElementById('roomheader').innerHTML = name;
};

const copyArray = function(xs, ys) {
    xs.length = 0;
    ys.forEach(x => xs.push(x));
};

const copyObj = function(xs, ys) {
    Object.keys(ys).forEach(k => {
        xs[k] = ys[k];
    });
    Object.keys(xs).forEach(k => {
        if (!ys.hasOwnProperty(k)) {
            delete xs[k];
        }
    });
};

const updateWorld = function(msg) {
    copyObj(attendees, msg['attendeeToCount']);
    copyArray(localStack, msg['local']);
    copyArray(globalStack, msg['broad']);
    updateAttendees(attendees);
    newQueue(LOCALSTACK, localStack);
    newQueue(BROADSTACK, globalStack);
    setRoomName(msg['name']);
};

const setUserDisplay = function(username) {
    document.getElementById('userdisplay').innerHTML = `Hello, ${username}!`;
};

const initRoom = function() {
    const initMsg = { action: "CLIENTINIT", room: roomid, attendee: username};
    debugLog(`Initializing with ${JSON.stringify(initMsg)}`);
    ws.send(JSON.stringify(initMsg));
};

const retryUser = function() {
    debugLog('Retrying user');
    username = '';
    while (username === '') {
        username = prompt("Your username is taken, please provide a new name");
    }
    initRoom();
    const exDate = new Date();
    exDate.setDate(exDate.getDate() + 1);
    document.cookie = `${cookieName}=${username}; ${exDate.toUTCString()}; path=/; SameSite=Strict`;
    setUserDisplay(username);
};

ws.onmessage = evt => {
    const m = JSON.parse(evt.data);
    debugLog(m);

    switch (m['action']) {
    case "DUPLICATEUSER":
        debugLog('Duplicate user');
        updateAttendees(m['attendees']);
        retryUser();
        break;
    case UPDATE_WORLD:
        debugLog('Updating entire page');
        updateWorld(m);
        break;
    default:
        console.log(`Unknown action: [${evt.data}]`);
    }
};

ws.onopen = initRoom;

const deleteCookie = function() {
    document.cookie = `${cookieName}=NA; expires=Fri, 14 May 2021 00:00:00 UTC; path=/; SameSite=Strict`;
};


const leaveRoom = function() {
    const leaveMsg = { action: "LEAVE" };
    ws.send(JSON.stringify(leaveMsg));
    ws.close();
    deleteCookie();
    window.location = `/?room=${roomid}`;
};

const closeRoom = function() {
    if(confirm("Are you sure? This will close the room for everyone!")) {
        const closeMsg = { action: "CLOSE" };
        ws.send(JSON.stringify(closeMsg));
        deleteCookie();
        window.location = '/';
    }
};

setUserDisplay(username);


static/invite.js
const verifyform = function(form) {
    let valid = true;
    Array.from(form.elements).forEach(el => {
        if (el.type === 'text' && !el.value) {
            alert(`${el.name} must be provided`);
            valid = false;
        }
    });
    return valid;
};

const copier = function(){
    const shareurl = document.getElementById('copylink').innerHTML;
    const el = document.createElement('textarea');
    el.value = shareurl;
    document.body.appendChild(el);
    el.select();
    el.setSelectionRange(0, 99999);
    document.execCommand('copy');
    document.body.removeChild(el);
    alert('Copied the link!');
};

const maybeRoomID = new URLSearchParams(location.search).get('room');

if (maybeRoomID) {
    document.getElementById('room-id').value = maybeRoomID;
}
if (typeof roomid !== 'undefined') {
    const shareurl = `http://${location.host}?room=${roomid}`;
    document.getElementById('copylink').innerHTML = shareurl;
};


Main.hs
{-# LANGUAGE NamedFieldPuns #-}
import           Protolude                            hiding (Text, local)
import           Protolude.Partial                    (read)

import           Data.Aeson
import           Data.Cache.LRU.IO                    (AtomicLRU)
import qualified Data.Cache.LRU.IO                    as Cache
import           Data.UUID                            (UUID)
import qualified Data.UUID                            as UUID
import qualified Data.UUID.V4                         as UUID
import           Data.String                          (fromString)
import qualified Data.Text                            as StrictText
import           Data.Text.Lazy                       (Text)
import qualified Data.Text.Lazy                       as Text
import           GHC.Base                             (String)
import qualified Network.HTTP.Types.Status            as Sc
import qualified Network.Wai.Middleware.Gzip          as Sc
import qualified Network.Wai.Handler.WebSockets       as WaiWs
import qualified Network.WebSockets                   as WS
import qualified Network.Wai                          as Wai
import qualified Network.Wai.Handler.Warp             as Warp
import qualified Network.Wai.Middleware.RequestLogger as Wai (logStdout)
import qualified System.Environment                   as Sys
import qualified Web.Scotty                           as Sc
import qualified Web.Scotty.Cookie                    as Sc

import           Data.Types

type RoomCache = AtomicLRU RoomID RoomState

data RoomPost = CreateRoom { roomNameParam :: Text, username :: Text }
              | JoinRoom   { roomIDParam :: RoomID, username :: Text }
              deriving (Eq, Ord, Read, Show)

userCookie :: UUID -> StrictText.Text
userCookie rid = "username-" <> (UUID.toText rid)

main :: IO ()
main = do
  port <- sysWithDefault "PORT" read 80
  host <- sysWithDefault "HOST" fromString "*4"
  keepAliveURL <- Sys.lookupEnv "KEEPALIVE_URL"
  let settings = Warp.setHost host $ Warp.setPort port Warp.defaultSettings
  cache <- Cache.newAtomicLRU Nothing
  sapp  <- scottyApp cache keepAliveURL
  let wsapp = websockApp cache
  Warp.runSettings settings $ WaiWs.websocketsOr WS.defaultConnectionOptions wsapp sapp

scottyApp :: RoomCache -> Maybe String -> IO Wai.Application
scottyApp cache keepAlive =
  Sc.scottyApp $ do
    Sc.middleware $ Sc.gzip (Sc.def { Sc.gzipFiles = Sc.GzipCompress }) . Wai.logStdout

    Sc.get "/" $
      Sc.file "index.html"
    Sc.post "/" $ do
      params <- Sc.params
      case mkRoomPost params of
        Nothing  -> Sc.status Sc.badRequest400
        Just req -> do
          case req of
            cr@CreateRoom{} -> createRoom cr cache
            jr@JoinRoom{}   -> do
              let rid = unRoomID . roomIDParam $ jr
              Sc.setSimpleCookie (userCookie rid) (Text.toStrict $ username req)
              Sc.redirect $ "/room/" <> uuidToText rid
    Sc.get "/static/cookie-policy/"  $ Sc.file "static/cookie-policy.html"
    Sc.get "/static/privacy-policy/" $ Sc.file "static/privacy.html"
    -- TODO: Apply some simple scheme to avoid spam creation of rooms
    Sc.get "/room/:pRoomID"    $ Sc.file "room.html"
    -- TODO: setup some static middleware
    Sc.get "/static/client.js" $ Sc.file "static/client.js"
    Sc.get "/static/invite.js" $ Sc.file "static/invite.js"
    Sc.get "/static/style.css" $ Sc.file "static/style.css"

    case keepAlive of
      Nothing  -> pure ()
      Just url -> do
        Sc.get (Sc.capture url) $ Sc.file "static/keepalive.html"
        Sc.get "/static/keepalive.js" $ Sc.file "static/keepalive.js"

websockApp :: RoomCache -> WS.ServerApp
websockApp cache pending = do
  putText "ws connected"
  conn <- WS.acceptRequest pending
  WS.forkPingThread conn 30 -- TODO: remove this deprecated dill

  maybeRooms <- joinRoom conn cache
  case maybeRooms of
    Nothing -> pure ()
    Just (initRoom, initRoom', CLIENTINIT{room, attendee}) -> do
      _ <- Cache.insert room initRoom' cache
      _ <- pushWorld room conn cache nilRoom

      _ <- createReader attendee room conn cache
      -- this intentionally uses initRoom instead of initRoom'
      -- to force the first push of the state of the room
      _ <- iterate_ initRoom $ \rstate -> do
        curState <- pushWorld room conn cache rstate
        threadDelay 100000
        pure curState
      pure ()
    Just _ -> WS.sendClose conn ("Error while joining room" :: Text)

joinRoom :: WS.Connection -> RoomCache -> IO (Maybe (RoomState, RoomState, Payload))
joinRoom conn cache = do
  roomidReq <- WS.receiveData conn
  case decode roomidReq of
    Nothing -> WS.sendClose conn ("Bad UUID byyeeeeeeee" :: Text) >> pure Nothing
    Just cli@CLIENTINIT{room, attendee} -> do
      maybeRoom <- Cache.lookup room cache
      case maybeRoom of
        Nothing       -> WS.sendClose conn ("byyyyyyyeeeee" :: Text) >> pure Nothing
        Just initRoom -> do
          case newUser attendee initRoom of
            Just initRoom' -> pure $ pure (initRoom, initRoom', cli)
            Nothing        -> do
              _ <- WS.sendTextData conn . encode $ DUPLICATEUSER (userList initRoom)
              joinRoom conn cache
    Just _ -> WS.sendClose conn ("Bad payload byyyeee" :: Text) >> pure Nothing

mkRoomPost :: [Sc.Param] -> Maybe RoomPost
mkRoomPost params = maybeJoin <|> maybeCreate
  where
  maybeJoin    = do
    roomIDP  <- param "room-id" params
    rID      <- textToRoomID roomIDP
    username <- param "username" params
    pure $ JoinRoom rID username
  maybeCreate  = do
    roomName  <- param "room-name" params
    username  <- param "username" params
    pure $ CreateRoom roomName username

createRoom :: RoomPost -> RoomCache -> Sc.ActionM ()
createRoom req cache = do
  rstate <- Sc.liftAndCatchIO $ do
    rID   <- fmap RoomID UUID.nextRandom
    let rstate = newRoom rID . roomNameParam $ req
    Cache.insert rID rstate cache
    putLText $ "Created room with ID [" <> roomIDToText rID <> "]"
    pure rstate

  let rid   = unRoomID $ roomID rstate
  Sc.setSimpleCookie (userCookie rid) (Text.toStrict $ username req)
  Sc.redirect $ "/room/" <> uuidToText rid

param :: Text -> [Sc.Param] -> Maybe Text
param var = fmap snd . find (\(par, _) -> par == var)

pushWorld :: RoomID -> WS.Connection -> RoomCache -> RoomState -> IO RoomState
pushWorld rid conn cache prevState = do
  maybeRoom <- Cache.lookup rid cache
  case maybeRoom of
    Nothing   -> WS.sendClose conn ("byyyyyyyeeeee" :: Text) >> pure prevState
    Just room -> do
      when (room /= prevState) $ WS.sendTextData conn . encode $ WORLD {
        attendeeToCount = users room,
        local           = toList (rlocal  room),
        broad           = toList (rglobal room),
        name            = roomName room
      }
      pure room

createReader :: UserID -> RoomID -> WS.Connection -> RoomCache -> IO ThreadId
createReader attendee rid conn cache = forkIO . forever $ do
  msg       <- WS.receiveData conn
  maybeRoom <- Cache.lookup rid cache
  case maybeRoom of
    Nothing       -> WS.sendClose conn ("Room is closed byyyyee" :: Text)
    Just room -> do
      case decode msg of
        Nothing -> WS.sendClose conn ("Bad UUID byyeeeeeeee" :: Text)
        Just QUEUE{stack} -> do
            let room' = q stack attendee room
            Cache.insert rid room' cache
        Just QOTHER{stack, other} -> do
            let room' = q stack other room
            Cache.insert rid room' cache
        Just NEXT{stack} -> do
          case stack of
            LOCAL -> Cache.insert rid (nextLocal room) cache
            BROAD ->
              if lengthLocal room > 1
              then pure () -- Do nothing, don't delete an incomplete local stack
              else case peakNextGlobal room of
                     Nothing          -> pure () -- Nothing in global to mess with
                     Just nextSpeaker -> do
                       let room' = qLocal nextSpeaker . nextLocal . nextGlobal $ room
                       Cache.insert rid room' cache
        Just REORDER{stack, newstack} -> do
          let room' = reorder newstack stack room
          Cache.insert rid room' cache
        Just CANCEL{stack, newstack}  -> do
          let room' = cancelStack newstack stack room
          Cache.insert rid room' cache
        Just ADD{other} -> do
          case newUser other room of
            Nothing    -> pure ()
            Just room' -> Cache.insert rid room' cache
        Just LEAVE       -> do
          let room' = deleteUser attendee room
          Cache.insert rid room' cache
          WS.sendClose conn ("Thanks for joining!" :: Text)
        Just CLOSE       -> do
          _ <- Cache.delete rid cache
          WS.sendClose conn ("Thanks for joining!" :: Text)
        _       -> WS.sendClose conn ("Unsupported byyyyeee" :: Text)

iterate_ :: Monad f => a -> (a -> f a) -> f a
iterate_ a f = f a >>= \a' -> iterate_ a' f

sysWithDefault :: String -> (String -> a) -> a -> IO a
sysWithDefault var f x = fmap (fromMaybe x . fmap f) $ Sys.lookupEnv var
